File: src/main/kotlin/Day20B.go
package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

type Node struct {
	value int64
	pos   int
	left  *Node
	right *Node
}

func move(n *Node, c int) {
	for c > 0 {
		myOldLeft := n.left

		other := n.right
		otherRight := n.right.right

		n.left = other
		n.right = otherRight

		other.left = myOldLeft
		other.right = n

		myOldLeft.right = other
		otherRight.left = n

		myPos := n.pos
		n.pos = other.pos
		other.pos = myPos

		c--
	}

	for c < 0 {
		myOldRight := n.right

		other := n.left
		otherLeft := n.left.left

		n.left = otherLeft
		n.right = other

		other.left = n
		other.right = myOldRight

		myOldRight.left = other
		otherLeft.right = n

		myPos := n.pos
		n.pos = other.pos
		other.pos = myPos

		c++
	}
}

func findVal(n *Node, v int64) *Node {
	for n.value != v {
		n = n.right
	}

	return n
}

func withOffset(n *Node, offset int) *Node {
	i := 0
	for i < offset {
		n = n.right
		i++
	}

	return n
}

func main() {
	strs := readLines("/Users/alekseik/software/adventofcode2022/data/day20.txt")
	nums := toNums(strs)

	size := len(nums)
	elements := make([]Node, size)
	var prev *Node

	for i, v := range nums {
		elements[i] = Node{v * 811589153, i, nil, nil}

		if prev != nil {
			prev.right = &elements[i]
			(&elements[i]).left = prev
		}
		prev = &elements[i]
	}
	prev.right = &elements[0]
	elements[0].left = prev

	queue := make([]*Node, size*10)
	for i := 0; i < 10; i++ {
		for j, v := range elements {
			queue[j+(i*size)] = &v
		}
	}

	loop := int64((size) * (size - 1))

	for i, v := range queue {
		move(v, int(v.value%loop))
		println("Iteration", i, "of total", len(queue))
	}

	zero := findVal(queue[0], 0)
	var result int64 = 0
	for i := 1000; i <= 3000; i += 1000 {
		value := withOffset(zero, i).value
		println("Value", value)
		result += value
	}

	println("Result", result)
}

func toNums(s []string) []int64 {
	c := make([]int64, len(s))
	for i, v := range s {
		iv, e := strconv.ParseInt(v, 0, 64)
		if e != nil {
			panic(e)
		}

		c[i] = iv
	}

	return c
}

func readLines(path string) []string {
	file, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := scanner.Text()
		s = strings.TrimSpace(s)

		if len(s) > 0 {
			lines = append(lines, s)
		}
	}

	return lines
}

File: src/main/kotlin/Day20B.kt
import java.util.LinkedList

object Day20B {

    var X = ArrayList<Node>()

    fun printList() {
        X.forEach {
            print("${it.v}, ")
        }
        println()
    }

    data class Node(
        val v: Long,
        var pos: Int
    ) {
        fun move(pc: Int) {
            var c = pc
            while (c != 0) {
                val myPos = pos
                val otherPos: Int

                if (c > 0) {
                    otherPos = rightPos()
                    c--

                } else {
                    otherPos = leftPos()
                    c++
                }

                if (myPos == X.size - 1 && otherPos == 0) {
                    var prev: Node = this
                    for (i in 0 until X.size ) {
                        val curr = X[i]
                        X[i] = prev
                        X[i].pos = i

                        prev = curr
                    }
                    c++
                } else if (myPos == 0 && otherPos == X.size - 1) {
                    for (i in 1 until X.size ) {
                        val curr = X[i]
                        X[i - 1] = curr
                        X[i - 1].pos = i - 1
                    }
                    X[X.size - 1] = this
                    pos = X.size - 1
                    c--
                } else {
                    val otherNode = X[otherPos]

                    X[otherPos] = this
                    this.pos = otherPos

                    X[myPos] = otherNode
                    otherNode.pos = myPos
                }
            }
        }

        private fun leftPos() = if (pos - 1 >= 0) pos - 1 else X.size - 1

        private fun rightPos() = (pos + 1) % X.size

        fun findVal(t: Long): Node {
            var current = this
            while (current.v != t) {
                current = X[current.rightPos()]
            }

            return current
        }

        fun withOffset(offset: Long): Node {
            var current = this
            for (i in 0 until offset) {
                current = X[current.rightPos()]
            }

            return current
        }

        override fun toString(): String {
            return "Node(v=$v, pos=$pos)"
        }

    }

    fun run() {
        val nums = "day20.txt"
            .readFile()
            .lines()
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .map { it.toLong() }
            .map { it * 811589153  }

        val queue = LinkedList<Node>()
        val size = nums.size.toLong()

        val root = Node(nums[0], 0)
        queue.add(root)
        X.add(root)

        for (i in 1 until nums.size) {
            val current = Node(nums[i], i)
            queue.add(current)
            X.add(current)
        }

        val queueCopy = ArrayList(queue)
        repeat(9) {
            queue.addAll(queueCopy)
        }

        val loop: Long = (size - 1)
        printList()

        while (queue.isNotEmpty()) {
            val el = queue.removeFirst()
            el.move((el.v % loop).toInt())
//            print("Move ${el.v}: ")
//            printList()
            println("Queue size = ${queue.size}")
        }

        val zero = root.findVal(0)
        var result = 0L
        for (i in listOf(1000, 2000, 3000)) {
            val v = zero.withOffset(i.toLong()).v
            println("V=$v")
            result += v
        }

        /*
        V=6801
        V=9840
        V=849
        Res = 17490
         */
        println("Res = $result")
    }

}

fun main() {
    Day20B.run()
}
